### ISSUE TITLE
Incorrect `needless_range_loop` warning with a 2-dimensional array

### ISSUE THREAD

**Issue by pubfnbar** [rust-lang/rust-clippy#17670]
### Summary

Clippy doesn't understand that a variable is not in scope at the line where it suggests it should be used.

### Lint Name

clippy::needless_range_loop

### Reproducer

I tried this code:

```rust
fn main() {
    let mut arr2d = [[0; 3]; 2];

    for j in 0..3 {
        for i in 0..2 {
            arr2d[i][j] = i * j;
        }
    }
    println!("{arr2d:?}");
}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: the loop variable `j` is used to index `arr2d[i]`
 --> src/main.rs:4:14
  |
4 |     for j in 0..3 {
  |              ^^^^
  |
note: for this index operation
 --> src/main.rs:6:13
  |
6 |             arr2d[i][j] = i * j;
  |             ^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/main/index.html#needless_range_loop
  = note: `#[warn(clippy::needless_range_loop)]` on by default
help: consider using an iterator and `.enumerate()`
  |
4 -     for j in 0..3 {
4 +     for (j, <item>) in arr2d[i].iter_mut().enumerate() {
  |

warning: the loop variable `i` is used to index `arr2d`
 --> src/main.rs:5:18
  |
5 |         for i in 0..2 {
  |                  ^^^^
  |
note: for this index operation
 --> src/main.rs:6:13
  |
6 |             arr2d[i][j] = i * j;
  |             ^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/main/index.html#needless_range_loop
help: consider using an iterator and `.enumerate()`
  |
5 -         for i in 0..2 {
5 +         for (i, <item>) in arr2d.iter_mut().enumerate() {
  |
```

I expected to see this happen:
I expected no warning.

### Version

```text

```

**Comment by Halloloid** [comment]
I think that for cases like this, it may be better not to trigger `needless_range_loop`.

Here, both `i` and `j` are not only used for indexing, but are also part of the actual computation (`i * j`). Rewriting this using iterators and `enumerate()` makes the code more complicated, and the current suggestion for the outer loop is also invalid because `i` is not in scope there.

Perhaps the lint should avoid suggesting a rewrite when the index is meaningfully used beyond indexing.

**Comment by pubfnbar** [comment]
I think I should have used the following example code to make it so that the two loops couldn't be swapped, making it impossible to rewrite it by using iterators, `enumerate()`, and simply changing the order of iteration over `arr2d`:
```rust
fn main() {
    let mut arr2d = [[0; 3]; 2];
    let mut x = 0;

    for j in 0..3 {
        for i in 0..2 {
            arr2d[i][j] = x;
            x += 1;
        }
    }
    println!("{arr2d:?}");
}
```

**Comment by 0xull** [comment]
hello @pubfnbar & @Halloloid i would love to attempt this. still digging into the code site, but i thought to say 'hi' whilst i read the rest `needless_range_loop`.

**Comment by pubfnbar** [comment]
@0xull I haven't looked into it, but this issue might be a duplicate of [#17432](https://github.com/rust-lang/rust-clippy/issues/17432).

**Comment by 0xull** [comment]
@pubfnbar ohh righttt. it looks the same. thanks.

**Comment by CommanderStorm** [comment]
@rustbot close

Agree, looks like a dup of https://github.com/rust-lang/rust-clippy/issues/17432 and there is an open PR
- https://github.com/rust-lang/rust-clippy/pull/17438
